pkg/cloud_provider/user_tags.go

```diff
@@ -0,0 +1,536 @@
+	"time"
+
+	rscmgr "cloud.google.com/go/resourcemanager/apiv3"
+	rscmgrpb "cloud.google.com/go/resourcemanager/apiv3/resourcemanagerpb"
+	"github.com/googleapis/gax-go/v2"
+	"github.com/googleapis/gax-go/v2/apierror"
+	"golang.org/x/time/rate"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+	"k8s.io/klog/v2"
+)
+
+const (
+	// ParameterKeyResourceTags is the name of the key defined in the parameters'
+	// field of the StorageClass object, which holds the user defined resource tags.
+	ParameterKeyResourceTags = "resource-tags"
+
+	// tagsDelimiter is the character for separating the tag of the form
+	// <parentID>/<tagKey>/<tagValue> in to parentID, tag key and tag value.
+	tagsDelimiter = "/"
+
+	// maxTagsPerResource is the maximum number of resource tags that can
+	// be attached to a resource.
+	// https://cloud.google.com/resource-manager/docs/limits#tag-limits
+	maxTagsPerResource = 50
+
+	// resourceManagerHostSubPath is the endpoint for tag requests.
+	resourceManagerHostSubPath = "cloudresourcemanager.googleapis.com"
+
+	// filestoreInstanceFullNameFmt is the string format for the full resource name of a
+	// filestore Instance resource.
+	filestoreInstanceFullNameFmt = "//file.googleapis.com/projects/%s/locations/%s/instances/%s"
+
+	// filestoreBackupFullNameFmt is the string format for the full resource name of a
+	// filestore Backup resource.
+	filestoreBackupFullNameFmt = "//file.googleapis.com/projects/%s/locations/%s/backups/%s"
+
+	// tagsRequestRateLimit is the tag request rate limit per second.
+	tagsRequestRateLimit = 8
+
+	// tagsRequestTokenBucketSize is the burst/token bucket size used.
+	// for limiting API requests.
+	tagsRequestTokenBucketSize = 8
+)
+
+const (
+	// FilestoreInstance is for specifying GCP Filestore Instance resource.
+	FilestoreInstance resourceType = "FilestoreInstance"
+
+	// FilestoreBackUp is for specifying GCP Filestore Backup resource.
+	FilestoreBackUp resourceType = "FilestoreBackUp"
+)
+
+var (
+	// resourceTagsValueFiller is the filler value for resourceTags type.
+	resourceTagsValueFiller = struct{}{}
+)
+
+// resourceType is for identifying the GCP resource.
+type resourceType string
+
+// resourceTags is the custom type for holding tags.
+type resourceTags map[string]struct{}
+
+// withoutAuthentication is for indicating to set WithoutAuthentication client
+// option when creating GCP tag client. Setting this will result in no other
+// authentication options being used.
+type withoutAuthentication struct{}
+
+// tagServiceManager handles resource tagging.
+type tagServiceManager struct {
+	*Cloud
+	tags resourceTags
+
+	// httpClient is currently used for mocking client for tests.
+	httpClient *http.Client
+	// httpEndpoint is the endpoint used for tests and can be used
+	// with httpClient only.
+	httpEndpoint string
+	// withoutAuthentication is currently used only for tests. It can
+	// also be utilized for accessing public resources.
+	withoutAuthentication bool
```

Comment:
This is a new change for which a PR is created upstream and yet to be merged at this moment. This was required to remove creds requirement in UTs.

Reply:
It is not merged right?

Reply:
No, https://github.com/kubernetes-sigs/gcp-filestore-csi-driver/pull/905 is created with the same changes. Upstream didn't require this change since the CI environment had the creds(UTs should not dependent on it) and hence UTs were going through there. I am following it up to get it merged in the upstream.